{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Butterfly Analyses\n",
    "## Dependencies\n",
    "First we will load the necessary packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages.\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import norm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we define a function to calculate site patterns from derived allele frequency arrays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def butterfly_site_patterns(\n",
    "    p1,\n",
    "    p2,\n",
    "    p3,\n",
    "):\n",
    "    \"\"\"\n",
    "    ###########################################################################\n",
    "    INPUT\n",
    "        p1: Derived alle frequency array for the P1 lineage.\n",
    "        p2: Derived alle frequency array for the P2 lineage.\n",
    "        p3: Derived alle frequency array for the P3 lineage.\n",
    "    ---------------------------------------------------------------------------\n",
    "    OUTPUT: Genome counts of ABBA, BABA, BBAA, BAAA, ABAA, and AABA sites.\n",
    "    ###########################################################################\n",
    "    \"\"\"\n",
    "    # Calculate site pattern counts.\n",
    "    abba = ((1 - p1) * (p2) * (p3)).sum()\n",
    "    baba = ((p1) * (1 - p2) * (p3)).sum()\n",
    "    baaa = ((p1) * (1 - p2) * (1 - p3)).sum()\n",
    "    abaa = ((1 - p1) * (p2) * (1 - p3)).sum()\n",
    "    return abba, baba, baaa, abaa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we define a function to calculate all the introgression metrics for detecting introgression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def butterfly_detection(\n",
    "    abba,\n",
    "    baba,\n",
    "    baaa,\n",
    "    abaa,\n",
    "):\n",
    "    \"\"\"\n",
    "    ###########################################################################\n",
    "    INPUT: Genome-wide ABBA, BABA, BAAA, and ABAA counts from frequencies.\n",
    "    ---------------------------------------------------------------------------\n",
    "    OUTPUT: Patterson's D, Danc, and D+ values.\n",
    "    ###########################################################################\n",
    "    \"\"\"\n",
    "    # Calculate Patterson's D.\n",
    "    d = ((abba - baba) / (abba + baba))\n",
    "    # Calculate Danc.\n",
    "    danc = ((baaa - abaa) / (baaa + abaa))\n",
    "    # Calculate D+.\n",
    "    dplus = (((abba - baba) + (baaa - abaa)) / ((abba + baba) + (baaa + abaa)))\n",
    "    return d, danc, dplus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we define a function to calculate all the introgression metrics for quantifying introgression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def butterfly_quantification(\n",
    "    abba_num,\n",
    "    baba_num,\n",
    "    baaa_num,\n",
    "    abaa_num,\n",
    "    abba_den,\n",
    "    baba_den,\n",
    "    baaa_den,\n",
    "    abaa_den,\n",
    "):\n",
    "    \"\"\"\n",
    "    ###########################################################################\n",
    "    INPUT\n",
    "        {site pattern}_num: Site pattern counts from ((P1, P2), P3).\n",
    "        {site pattern}_den: Site pattern counts from ((P1, P3), P3).\n",
    "    ---------------------------------------------------------------------------\n",
    "    OUTPUT: fhom, fanc, and f+ values.\n",
    "    ###########################################################################\n",
    "    \"\"\"\n",
    "    # Calculate fhom.\n",
    "    fhom = ((abba_num - baba_num) / (abba_den - baba_den))\n",
    "    # Calculate fanc.\n",
    "    fanc = ((baaa_num - abaa_num) / (baaa_den - abaa_den))\n",
    "    # Calculate f+.\n",
    "    fplus = (((abba_num - baba_num) + (baaa_num - abaa_num)) / ((abba_den - baba_den) + (baaa_den - abaa_den)))\n",
    "    return fhom, fanc, fplus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we define a function standard deviations of each bootstrapped distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_bs_stds(\n",
    "    p1,\n",
    "    p2,\n",
    "    p3,\n",
    "):\n",
    "    \"\"\"\n",
    "    ###########################################################################\n",
    "    INPUT\n",
    "        p1: P1 lineage shorthand.\n",
    "        p2: P2 lineage shorthand.\n",
    "        p3: P3 lineage shorthand.\n",
    "    ---------------------------------------------------------------------------\n",
    "    OUTPUT: Dictionary of standard deviations per bootstrapped distribution.\n",
    "    ###########################################################################\n",
    "    \"\"\"\n",
    "    # Define the file path for the results.\n",
    "    results_path = './bootstraps/{0}_{1}_{2}/'.format(p1, p2, p3)\n",
    "    # Load the bootstrapped distributions.\n",
    "    bs_abba = np.loadtxt(\n",
    "        results_path+'abba.csv.gz',\n",
    "        delimiter=',',\n",
    "    )\n",
    "    bs_baba = np.loadtxt(\n",
    "        results_path+'baba.csv.gz',\n",
    "        delimiter=',',\n",
    "    )\n",
    "    bs_baaa = np.loadtxt(\n",
    "        results_path+'baaa.csv.gz',\n",
    "        delimiter=',',\n",
    "    )\n",
    "    bs_abaa = np.loadtxt(\n",
    "        results_path+'abaa.csv.gz',\n",
    "        delimiter=',',\n",
    "    )\n",
    "    bs_abba_hom = np.loadtxt(\n",
    "        results_path+'abba_hom.csv.gz',\n",
    "        delimiter=',',\n",
    "    )\n",
    "    bs_baba_hom = np.loadtxt(\n",
    "        results_path+'baba_hom.csv.gz',\n",
    "        delimiter=',',\n",
    "    )\n",
    "    bs_baaa_hom = np.loadtxt(\n",
    "        results_path+'baaa_hom.csv.gz',\n",
    "        delimiter=',',\n",
    "    )\n",
    "    bs_abaa_hom = np.loadtxt(\n",
    "        results_path+'abaa_hom.csv.gz',\n",
    "        delimiter=',',\n",
    "    )\n",
    "    # Calculate bootstrapped introgression values.\n",
    "    bs_d, bs_danc, bs_dplus = butterfly_detection(\n",
    "        abba=bs_abba,\n",
    "        baba=bs_baba,\n",
    "        baaa=bs_baaa,\n",
    "        abaa=bs_abaa,\n",
    "    )\n",
    "    bs_fhom, bs_fanc, bs_fplus = butterfly_quantification(\n",
    "        abba_num=bs_abba,\n",
    "        baba_num=bs_baba,\n",
    "        baaa_num=bs_baaa,\n",
    "        abaa_num=bs_abaa,\n",
    "        abba_den=bs_abba_hom,\n",
    "        baba_den=bs_baba_hom,\n",
    "        baaa_den=bs_baaa_hom,\n",
    "        abaa_den=bs_abaa_hom,\n",
    "    )\n",
    "    # Create a dictionary where key is the introgression metric and its\n",
    "    # corresponding numpy array is the value.\n",
    "    bs_std_dict = {\n",
    "        'd': np.std(bs_d), 'danc': np.std(bs_danc), 'dplus': np.std(bs_dplus),\n",
    "        'fhom': np.std(bs_fhom), 'fanc': np.std(bs_fanc), 'fplus': np.std(bs_fplus),\n",
    "    }\n",
    "    return bs_std_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a function to calculate p-values for introgression detection metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_p_values(\n",
    "    obs_d,\n",
    "    obs_danc,\n",
    "    obs_dplus,\n",
    "    bs_std_d,\n",
    "    bs_std_danc,\n",
    "    bs_std_dplus,\n",
    "):\n",
    "    \"\"\"\n",
    "    ###########################################################################\n",
    "    INPUT\n",
    "        obs_{stat}: Observed introgression value.\n",
    "        bs_std_{stat}: Standard deviation of the bootstrapped distribution.\n",
    "    ---------------------------------------------------------------------------\n",
    "    OUTPUT: P-values for introgression detection metrics.\n",
    "    ###########################################################################\n",
    "    \"\"\"\n",
    "    # Use the survival function to calculate p-values.\n",
    "    d_pval = norm.sf(x=abs(obs_d), loc=0, scale=abs(bs_std_d))\n",
    "    danc_pval = norm.sf(x=abs(obs_danc), loc=0, scale=abs(bs_std_danc))\n",
    "    dplus_pval = norm.sf(x=abs(obs_dplus), loc=0, scale=abs(bs_std_dplus))\n",
    "    # Create a dictionary where key is the introgression metric and its\n",
    "    # corresponding numpy array is the value.\n",
    "    pval_dict = {\n",
    "        'd': d_pval, 'danc': danc_pval, 'dplus': dplus_pval,\n",
    "    }\n",
    "    return pval_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we load the derived allele frequencies csv into a pandas dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>chrom_pos</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>position</th>\n",
       "      <th>chi</th>\n",
       "      <th>chi_1</th>\n",
       "      <th>chi_2</th>\n",
       "      <th>ros</th>\n",
       "      <th>ros_1</th>\n",
       "      <th>ros_2</th>\n",
       "      <th>...</th>\n",
       "      <th>am</th>\n",
       "      <th>am_1</th>\n",
       "      <th>am_2</th>\n",
       "      <th>ag</th>\n",
       "      <th>ag_1</th>\n",
       "      <th>ag_2</th>\n",
       "      <th>timP</th>\n",
       "      <th>timP_1</th>\n",
       "      <th>timP_2</th>\n",
       "      <th>silvaniforms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5609</td>\n",
       "      <td>HE670803</td>\n",
       "      <td>5608</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5613</td>\n",
       "      <td>HE670803</td>\n",
       "      <td>5612</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5642</td>\n",
       "      <td>HE670803</td>\n",
       "      <td>5641</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5643</td>\n",
       "      <td>HE670803</td>\n",
       "      <td>5642</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.875</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.875</td>\n",
       "      <td>0.75</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5664</td>\n",
       "      <td>HE670803</td>\n",
       "      <td>5663</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4783709</th>\n",
       "      <td>chrZ</td>\n",
       "      <td>4120681</td>\n",
       "      <td>HE671531</td>\n",
       "      <td>172429</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4783710</th>\n",
       "      <td>chrZ</td>\n",
       "      <td>4120692</td>\n",
       "      <td>HE671531</td>\n",
       "      <td>172440</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4783711</th>\n",
       "      <td>chrZ</td>\n",
       "      <td>4120701</td>\n",
       "      <td>HE671531</td>\n",
       "      <td>172449</td>\n",
       "      <td>0.625</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4783712</th>\n",
       "      <td>chrZ</td>\n",
       "      <td>4120719</td>\n",
       "      <td>HE671531</td>\n",
       "      <td>172467</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4783713</th>\n",
       "      <td>chrZ</td>\n",
       "      <td>4120735</td>\n",
       "      <td>HE671531</td>\n",
       "      <td>172483</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4783714 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome  chrom_pos  scaffold  position    chi  chi_1  chi_2    ros  \\\n",
       "0             chr1       5609  HE670803      5608  0.125   0.25   0.00  0.000   \n",
       "1             chr1       5613  HE670803      5612  0.125   0.25   0.00  0.000   \n",
       "2             chr1       5642  HE670803      5641  0.125   0.25   0.00  0.000   \n",
       "3             chr1       5643  HE670803      5642  0.000   0.00   0.00  0.875   \n",
       "4             chr1       5664  HE670803      5663  1.000   1.00   1.00  0.125   \n",
       "...            ...        ...       ...       ...    ...    ...    ...    ...   \n",
       "4783709       chrZ    4120681  HE671531    172429  1.000   1.00   1.00  1.000   \n",
       "4783710       chrZ    4120692  HE671531    172440  0.375   0.50   0.25  1.000   \n",
       "4783711       chrZ    4120701  HE671531    172449  0.625   0.50   0.75  0.000   \n",
       "4783712       chrZ    4120719  HE671531    172467  0.000   0.00   0.00  0.000   \n",
       "4783713       chrZ    4120735  HE671531    172483  1.000   1.00   1.00  1.000   \n",
       "\n",
       "         ros_1  ros_2  ...   am  am_1  am_2   ag  ag_1  ag_2   timP  timP_1  \\\n",
       "0          0.0   0.00  ...  0.0   0.0   0.0  0.0   0.0   0.0  0.000    0.00   \n",
       "1          0.0   0.00  ...  0.0   0.0   0.0  0.0   0.0   0.0  0.000    0.00   \n",
       "2          0.0   0.00  ...  0.0   0.0   0.0  0.0   0.0   0.0  0.000    0.00   \n",
       "3          1.0   0.75  ...  1.0   1.0   1.0  1.0   1.0   1.0  0.875    0.75   \n",
       "4          0.0   0.25  ...  0.0   0.0   0.0  0.0   0.0   0.0  0.125    0.25   \n",
       "...        ...    ...  ...  ...   ...   ...  ...   ...   ...    ...     ...   \n",
       "4783709    1.0   1.00  ...  1.0   1.0   1.0  1.0   1.0   1.0  1.000    1.00   \n",
       "4783710    1.0   1.00  ...  1.0   1.0   1.0  1.0   1.0   1.0  0.000    0.00   \n",
       "4783711    0.0   0.00  ...  0.0   0.0   0.0  0.0   0.0   0.0  0.375    0.50   \n",
       "4783712    0.0   0.00  ...  0.0   0.0   0.0  0.0   0.0   0.0  0.000    0.00   \n",
       "4783713    1.0   1.00  ...  1.0   1.0   1.0  1.0   1.0   1.0  1.000    1.00   \n",
       "\n",
       "         timP_2  silvaniforms  \n",
       "0          0.00             0  \n",
       "1          0.00             0  \n",
       "2          0.00             0  \n",
       "3          1.00             0  \n",
       "4          0.00             0  \n",
       "...         ...           ...  \n",
       "4783709    1.00             0  \n",
       "4783710    0.00             0  \n",
       "4783711    0.25             0  \n",
       "4783712    0.00             0  \n",
       "4783713    1.00             0  \n",
       "\n",
       "[4783714 rows x 24 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load in the butterfly derived allele frequencies.\n",
    "der_freq_df = pd.read_csv('./butterflies_wgs_filtered_der_freqs.csv')\n",
    "der_freq_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lastly, we extract the genome-wide derived allele frequencies for each population as numpy arrays to make computations easier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract derived allele frequency arrays.\n",
    "# H. m. aglaope (aglaope).\n",
    "aglaope_der_freq_array = der_freq_df['ag'].values\n",
    "# H. m. amaryllis (amaryllis).\n",
    "amaryllis_der_freq_array = der_freq_df['am'].values\n",
    "# H. t. thelxinoe (timareta).\n",
    "timareta_der_freq_array = der_freq_df['timP'].values\n",
    "# H. m. melpomene from French Guianna (melp_fg).\n",
    "melp_fg_der_freq_array = der_freq_df['melFG'].values\n",
    "# H. m. melpomene from Panama (melp_p).\n",
    "melp_p_der_freq_array = der_freq_df['melP'].values\n",
    "# H. m. rosina (rosina).\n",
    "rosina_der_freq_array = der_freq_df['ros'].values\n",
    "# H. c. chioneus (cydno).\n",
    "cydno_der_freq_array = der_freq_df['chi'].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introgression Between  _H. timareta_ & _H. m. amaryllis_\n",
    "First, we re-examined rates of gene flow between _H. timareta_ and _H. m. amaryllis_ across three time periods: a short, recent period, subsequent to the divergence between _H. m. amaryllis_ and _H. m. aglaope_; an intermediate period, subsequent to the divergence of the Peruvian populations from French Guianan _H. m. melpomene_; and a long period, subsequent to the divergence between Peruvian and Panamanian populations.  \n",
    "  \n",
    "First, we will look at the short time period 4—i.e., ((_H. m. aglaope_, _H. m. amaryllis_), _H. timareta_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary for time period 4 ((H. m. aglaope, H. m. amaryllis), H. timareta)\n",
      "*** Site Pattern Counts ***\n",
      "ABBA - BABA: 2413.309\n",
      "BAAA - ABAA: 959.368\n",
      "*** Detection Metrics ***\n",
      "D - p-value: 0.039 - 0.0\n",
      "Danc - p-value: 0.003 - 0.008\n",
      "D+ - p-value: 0.008 - 0.0\n",
      "*** Quantification Metrics ***\n",
      "fhom - std: 0.018 - 0.003\n",
      "fanc - std: 0.008 - 0.003\n",
      "f+ - std: 0.013 - 0.003\n"
     ]
    }
   ],
   "source": [
    "# Calculate site pattern counts for.\n",
    "ag_am_tim_abba, ag_am_tim_baba, ag_am_tim_baaa, ag_am_tim_abaa = butterfly_site_patterns(\n",
    "    p1=aglaope_der_freq_array,\n",
    "    p2=amaryllis_der_freq_array,\n",
    "    p3=timareta_der_freq_array,\n",
    ")\n",
    "ag_tim_tim_abba, ag_tim_tim_baba, ag_tim_tim_baaa, ag_tim_tim_abaa = butterfly_site_patterns(\n",
    "    p1=aglaope_der_freq_array,\n",
    "    p2=timareta_der_freq_array,\n",
    "    p3=timareta_der_freq_array,\n",
    ")\n",
    "# Calculate site pattern differences.\n",
    "ag_am_tim_abba_baba = ag_am_tim_abba - ag_am_tim_baba\n",
    "ag_am_tim_baaa_abaa = ag_am_tim_baaa - ag_am_tim_abaa\n",
    "# Calculate introgression detection metrics.\n",
    "ag_am_tim_d, ag_am_tim_danc, ag_am_tim_dplus = butterfly_detection(\n",
    "    abba=ag_am_tim_abba,\n",
    "    baba=ag_am_tim_baba,\n",
    "    baaa=ag_am_tim_baaa,\n",
    "    abaa=ag_am_tim_abaa,\n",
    ")\n",
    "# Calculate introgression quantification metrics.\n",
    "ag_am_tim_fhom, ag_am_tim_fanc, ag_am_tim_fplus = butterfly_quantification(\n",
    "    abba_num=ag_am_tim_abba,\n",
    "    baba_num=ag_am_tim_baba,\n",
    "    baaa_num=ag_am_tim_baaa,\n",
    "    abaa_num=ag_am_tim_abaa,\n",
    "    abba_den=ag_tim_tim_abba,\n",
    "    baba_den=ag_tim_tim_baba,\n",
    "    baaa_den=ag_tim_tim_baaa,\n",
    "    abaa_den=ag_tim_tim_abaa,\n",
    ")\n",
    "# Load in the bootstrapped distributions.\n",
    "ag_am_tim_bs_dict = load_bs_stds(\n",
    "    p1='ag',\n",
    "    p2='am',\n",
    "    p3='timP',\n",
    ")\n",
    "# Calculate p-values for introgression detection method.\n",
    "ag_am_tim_pval_dict = calc_p_values(\n",
    "    obs_d=ag_am_tim_d,\n",
    "    obs_danc=ag_am_tim_danc,\n",
    "    obs_dplus=ag_am_tim_dplus,\n",
    "    bs_std_d=ag_am_tim_bs_dict['d'],\n",
    "    bs_std_danc=ag_am_tim_bs_dict['danc'],\n",
    "    bs_std_dplus=ag_am_tim_bs_dict['dplus'],\n",
    ")\n",
    "# Print a summary for this trio.\n",
    "print('Summary for time period 4 ((H. m. aglaope, H. m. amaryllis), H. timareta)')\n",
    "print('*** Site Pattern Counts ***')\n",
    "print('ABBA - BABA: {0}'.format(round(ag_am_tim_abba_baba, 3)))\n",
    "print('BAAA - ABAA: {0}'.format(round(ag_am_tim_baaa_abaa, 3)))\n",
    "print('*** Detection Metrics ***')\n",
    "print('D - p-value: {0} - {1}'.format(round(ag_am_tim_d, 3), round(ag_am_tim_pval_dict['d'], 3)))\n",
    "print('Danc - p-value: {0} - {1}'.format(round(ag_am_tim_danc, 3), round(ag_am_tim_pval_dict['danc'], 3)))\n",
    "print('D+ - p-value: {0} - {1}'.format(round(ag_am_tim_dplus, 3), round(ag_am_tim_pval_dict['dplus'], 3)))\n",
    "print('*** Quantification Metrics ***')\n",
    "print('fhom - std: {0} - {1}'.format(round(ag_am_tim_fhom, 3), round(ag_am_tim_bs_dict['fhom'], 3)))\n",
    "print('fanc - std: {0} - {1}'.format(round(ag_am_tim_fanc, 3), round(ag_am_tim_bs_dict['fanc'], 3)))\n",
    "print('f+ - std: {0} - {1}'.format(round(ag_am_tim_fplus, 3), round(ag_am_tim_bs_dict['fplus'], 3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will look at the intermediate time period 4 + 3—i.e., ((_H. m. melpomene (French Guianna)_, _H. m. amaryllis_), _H. timareta_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary for time period 4 + 3 ((H. m. melpomene (French Guianna), H. m. amaryllis), H. timareta)\n",
      "*** Site Pattern Counts ***\n",
      "ABBA - BABA: 14143.45\n",
      "BAAA - ABAA: 25927.758\n",
      "*** Detection Metrics ***\n",
      "D - p-value: 0.197 - 0.0\n",
      "Danc - p-value: 0.061 - 0.0\n",
      "D+ - p-value: 0.081 - 0.0\n",
      "*** Quantification Metrics ***\n",
      "fhom - std: 0.105 - 0.005\n",
      "fanc - std: 0.22 - 0.008\n",
      "f+ - std: 0.158 - 0.006\n"
     ]
    }
   ],
   "source": [
    "# Calculate site pattern counts for.\n",
    "melp_fg_am_tim_abba, melp_fg_am_tim_baba, melp_fg_am_tim_baaa, melp_fg_am_tim_abaa = butterfly_site_patterns(\n",
    "    p1=melp_fg_der_freq_array,\n",
    "    p2=amaryllis_der_freq_array,\n",
    "    p3=timareta_der_freq_array,\n",
    ")\n",
    "melp_fg_tim_tim_abba, melp_fg_tim_tim_baba, melp_fg_tim_tim_baaa, melp_fg_tim_tim_abaa = butterfly_site_patterns(\n",
    "    p1=melp_fg_der_freq_array,\n",
    "    p2=timareta_der_freq_array,\n",
    "    p3=timareta_der_freq_array,\n",
    ")\n",
    "# Calculate site pattern differences.\n",
    "melp_fg_am_tim_abba_baba = melp_fg_am_tim_abba - melp_fg_am_tim_baba\n",
    "melp_fg_am_tim_baaa_abaa = melp_fg_am_tim_baaa - melp_fg_am_tim_abaa\n",
    "# Calculate introgression detection metrics.\n",
    "melp_fg_am_tim_d, melp_fg_am_tim_danc, melp_fg_am_tim_dplus = butterfly_detection(\n",
    "    abba=melp_fg_am_tim_abba,\n",
    "    baba=melp_fg_am_tim_baba,\n",
    "    baaa=melp_fg_am_tim_baaa,\n",
    "    abaa=melp_fg_am_tim_abaa,\n",
    ")\n",
    "# Calculate introgression quantification metrics.\n",
    "melp_fg_am_tim_fhom, melp_fg_am_tim_fanc, melp_fg_am_tim_fplus = butterfly_quantification(\n",
    "    abba_num=melp_fg_am_tim_abba,\n",
    "    baba_num=melp_fg_am_tim_baba,\n",
    "    baaa_num=melp_fg_am_tim_baaa,\n",
    "    abaa_num=melp_fg_am_tim_abaa,\n",
    "    abba_den=ag_tim_tim_abba,\n",
    "    baba_den=ag_tim_tim_baba,\n",
    "    baaa_den=ag_tim_tim_baaa,\n",
    "    abaa_den=ag_tim_tim_abaa,\n",
    ")\n",
    "# Load in the bootstrapped distributions.\n",
    "melp_fg_am_tim_bs_dict = load_bs_stds(\n",
    "    p1='melFG',\n",
    "    p2='am',\n",
    "    p3='timP',\n",
    ")\n",
    "# Calculate p-values for introgression detection method.\n",
    "melp_fg_am_tim_pval_dict = calc_p_values(\n",
    "    obs_d=melp_fg_am_tim_d,\n",
    "    obs_danc=melp_fg_am_tim_danc,\n",
    "    obs_dplus=melp_fg_am_tim_dplus,\n",
    "    bs_std_d=melp_fg_am_tim_bs_dict['d'],\n",
    "    bs_std_danc=melp_fg_am_tim_bs_dict['danc'],\n",
    "    bs_std_dplus=melp_fg_am_tim_bs_dict['dplus'],\n",
    ")\n",
    "# Print a summary for this trio.\n",
    "print('Summary for time period 4 + 3 ((H. m. melpomene (French Guianna), H. m. amaryllis), H. timareta)')\n",
    "print('*** Site Pattern Counts ***')\n",
    "print('ABBA - BABA: {0}'.format(round(melp_fg_am_tim_abba_baba, 3)))\n",
    "print('BAAA - ABAA: {0}'.format(round(melp_fg_am_tim_baaa_abaa, 3)))\n",
    "print('*** Detection Metrics ***')\n",
    "print('D - p-value: {0} - {1}'.format(round(melp_fg_am_tim_d, 3), round(melp_fg_am_tim_pval_dict['d'], 3)))\n",
    "print('Danc - p-value: {0} - {1}'.format(round(melp_fg_am_tim_danc, 3), round(melp_fg_am_tim_pval_dict['danc'], 3)))\n",
    "print('D+ - p-value: {0} - {1}'.format(round(melp_fg_am_tim_dplus, 3), round(melp_fg_am_tim_pval_dict['dplus'], 3)))\n",
    "print('*** Quantification Metrics ***')\n",
    "print('fhom - std: {0} - {1}'.format(round(melp_fg_am_tim_fhom, 3), round(melp_fg_am_tim_bs_dict['fhom'], 3)))\n",
    "print('fanc - std: {0} - {1}'.format(round(melp_fg_am_tim_fanc, 3), round(melp_fg_am_tim_bs_dict['fanc'], 3)))\n",
    "print('f+ - std: {0} - {1}'.format(round(melp_fg_am_tim_fplus, 3), round(melp_fg_am_tim_bs_dict['fplus'], 3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will look at the long time period 4 + 3 + 2—i.e., ((_H. m. melpomene (Panama)_, _H. m. amaryllis_), _H. timareta_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary for time period 4 + 3 + 2 ((H. m. melpomene (Panama), H. m. amaryllis), H. timareta)\n",
      "*** Site Pattern Counts ***\n",
      "ABBA - BABA: 20634.37\n",
      "BAAA - ABAA: -21659.164\n",
      "*** Detection Metrics ***\n",
      "D - p-value: 0.229 - 0.0\n",
      "Danc - p-value: -0.051 - 0.0\n",
      "D+ - p-value: -0.002 - 0.334\n",
      "*** Quantification Metrics ***\n",
      "fhom - std: 0.153 - 0.01\n",
      "fanc - std: -0.184 - 0.013\n",
      "f+ - std: -0.004 - 0.009\n"
     ]
    }
   ],
   "source": [
    "# Calculate site pattern counts for.\n",
    "melp_p_am_tim_abba, melp_p_am_tim_baba, melp_p_am_tim_baaa, melp_p_am_tim_abaa = butterfly_site_patterns(\n",
    "    p1=melp_p_der_freq_array,\n",
    "    p2=amaryllis_der_freq_array,\n",
    "    p3=timareta_der_freq_array,\n",
    ")\n",
    "melp_p_tim_tim_abba, melp_p_tim_tim_baba, melp_p_tim_tim_baaa, melp_p_tim_tim_abaa = butterfly_site_patterns(\n",
    "    p1=melp_p_der_freq_array,\n",
    "    p2=timareta_der_freq_array,\n",
    "    p3=timareta_der_freq_array,\n",
    ")\n",
    "# Calculate site pattern differences.\n",
    "melp_p_am_tim_abba_baba = melp_p_am_tim_abba - melp_p_am_tim_baba\n",
    "melp_p_am_tim_baaa_abaa = melp_p_am_tim_baaa - melp_p_am_tim_abaa\n",
    "# Calculate introgression detection metrics.\n",
    "melp_p_am_tim_d, melp_p_am_tim_danc, melp_p_am_tim_dplus = butterfly_detection(\n",
    "    abba=melp_p_am_tim_abba,\n",
    "    baba=melp_p_am_tim_baba,\n",
    "    baaa=melp_p_am_tim_baaa,\n",
    "    abaa=melp_p_am_tim_abaa,\n",
    ")\n",
    "# Calculate introgression quantification metrics.\n",
    "melp_p_am_tim_fhom, melp_p_am_tim_fanc, melp_p_am_tim_fplus = butterfly_quantification(\n",
    "    abba_num=melp_p_am_tim_abba,\n",
    "    baba_num=melp_p_am_tim_baba,\n",
    "    baaa_num=melp_p_am_tim_baaa,\n",
    "    abaa_num=melp_p_am_tim_abaa,\n",
    "    abba_den=ag_tim_tim_abba,\n",
    "    baba_den=ag_tim_tim_baba,\n",
    "    baaa_den=ag_tim_tim_baaa,\n",
    "    abaa_den=ag_tim_tim_abaa,\n",
    ")\n",
    "# Load in the bootstrapped distributions.\n",
    "melp_p_am_tim_bs_dict = load_bs_stds(\n",
    "    p1='melP',\n",
    "    p2='am',\n",
    "    p3='timP',\n",
    ")\n",
    "# Calculate p-values for introgression detection method.\n",
    "melp_p_am_tim_pval_dict = calc_p_values(\n",
    "    obs_d=melp_p_am_tim_d,\n",
    "    obs_danc=melp_p_am_tim_danc,\n",
    "    obs_dplus=melp_p_am_tim_dplus,\n",
    "    bs_std_d=melp_p_am_tim_bs_dict['d'],\n",
    "    bs_std_danc=melp_p_am_tim_bs_dict['danc'],\n",
    "    bs_std_dplus=melp_p_am_tim_bs_dict['dplus'],\n",
    ")\n",
    "# Print a summary for this trio.\n",
    "print('Summary for time period 4 + 3 + 2 ((H. m. melpomene (Panama), H. m. amaryllis), H. timareta)')\n",
    "print('*** Site Pattern Counts ***')\n",
    "print('ABBA - BABA: {0}'.format(round(melp_p_am_tim_abba_baba, 3)))\n",
    "print('BAAA - ABAA: {0}'.format(round(melp_p_am_tim_baaa_abaa, 3)))\n",
    "print('*** Detection Metrics ***')\n",
    "print('D - p-value: {0} - {1}'.format(round(melp_p_am_tim_d, 3), round(melp_p_am_tim_pval_dict['d'], 3)))\n",
    "print('Danc - p-value: {0} - {1}'.format(round(melp_p_am_tim_danc, 3), round(melp_p_am_tim_pval_dict['danc'], 3)))\n",
    "print('D+ - p-value: {0} - {1}'.format(round(melp_p_am_tim_dplus, 3), round(melp_p_am_tim_pval_dict['dplus'], 3)))\n",
    "print('*** Quantification Metrics ***')\n",
    "print('fhom - std: {0} - {1}'.format(round(melp_p_am_tim_fhom, 3), round(melp_p_am_tim_bs_dict['fhom'], 3)))\n",
    "print('fanc - std: {0} - {1}'.format(round(melp_p_am_tim_fanc, 3), round(melp_p_am_tim_bs_dict['fanc'], 3)))\n",
    "print('f+ - std: {0} - {1}'.format(round(melp_p_am_tim_fplus, 3), round(melp_p_am_tim_bs_dict['fplus'], 3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will look at the long time period 4 + 3 + 2—i.e., ((_H. m. rosina_, _H. m. amaryllis_), _H. timareta_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary for time period 4 + 3 + 2 ((H. m. rosina, H. m. amaryllis), H. timareta)\n",
      "*** Site Pattern Counts ***\n",
      "ABBA - BABA: 19153.575\n",
      "BAAA - ABAA: -1221.242\n",
      "*** Detection Metrics ***\n",
      "D - p-value: 0.209 - 0.0\n",
      "Danc - p-value: -0.003 - 0.2\n",
      "D+ - p-value: 0.033 - 0.0\n",
      "*** Quantification Metrics ***\n",
      "fhom - std: 0.142 - 0.01\n",
      "fanc - std: -0.01 - 0.012\n",
      "f+ - std: 0.071 - 0.01\n"
     ]
    }
   ],
   "source": [
    "# Calculate site pattern counts for.\n",
    "rosina_am_tim_abba, rosina_am_tim_baba, rosina_am_tim_baaa, rosina_am_tim_abaa = butterfly_site_patterns(\n",
    "    p1=rosina_der_freq_array,\n",
    "    p2=amaryllis_der_freq_array,\n",
    "    p3=timareta_der_freq_array,\n",
    ")\n",
    "rosina_tim_tim_abba, rosina_tim_tim_baba, rosina_tim_tim_baaa, rosina_tim_tim_abaa = butterfly_site_patterns(\n",
    "    p1=rosina_der_freq_array,\n",
    "    p2=timareta_der_freq_array,\n",
    "    p3=timareta_der_freq_array,\n",
    ")\n",
    "# Calculate site pattern differences.\n",
    "rosina_am_tim_abba_baba = rosina_am_tim_abba - rosina_am_tim_baba\n",
    "rosina_am_tim_baaa_abaa = rosina_am_tim_baaa - rosina_am_tim_abaa\n",
    "# Calculate introgression detection metrics.\n",
    "rosina_am_tim_d, rosina_am_tim_danc, rosina_am_tim_dplus = butterfly_detection(\n",
    "    abba=rosina_am_tim_abba,\n",
    "    baba=rosina_am_tim_baba,\n",
    "    baaa=rosina_am_tim_baaa,\n",
    "    abaa=rosina_am_tim_abaa,\n",
    ")\n",
    "# Calculate introgression quantification metrics.\n",
    "rosina_am_tim_fhom, rosina_am_tim_fanc, rosina_am_tim_fplus = butterfly_quantification(\n",
    "    abba_num=rosina_am_tim_abba,\n",
    "    baba_num=rosina_am_tim_baba,\n",
    "    baaa_num=rosina_am_tim_baaa,\n",
    "    abaa_num=rosina_am_tim_abaa,\n",
    "    abba_den=ag_tim_tim_abba,\n",
    "    baba_den=ag_tim_tim_baba,\n",
    "    baaa_den=ag_tim_tim_baaa,\n",
    "    abaa_den=ag_tim_tim_abaa,\n",
    ")\n",
    "# Load in the bootstrapped distributions.\n",
    "rosina_am_tim_bs_dict = load_bs_stds(\n",
    "    p1='ros',\n",
    "    p2='am',\n",
    "    p3='timP',\n",
    ")\n",
    "# Calculate p-values for introgression detection method.\n",
    "rosina_am_tim_pval_dict = calc_p_values(\n",
    "    obs_d=rosina_am_tim_d,\n",
    "    obs_danc=rosina_am_tim_danc,\n",
    "    obs_dplus=rosina_am_tim_dplus,\n",
    "    bs_std_d=rosina_am_tim_bs_dict['d'],\n",
    "    bs_std_danc=rosina_am_tim_bs_dict['danc'],\n",
    "    bs_std_dplus=rosina_am_tim_bs_dict['dplus'],\n",
    ")\n",
    "# Print a summary for this trio.\n",
    "print('Summary for time period 4 + 3 + 2 ((H. m. rosina, H. m. amaryllis), H. timareta)')\n",
    "print('*** Site Pattern Counts ***')\n",
    "print('ABBA - BABA: {0}'.format(round(rosina_am_tim_abba_baba, 3)))\n",
    "print('BAAA - ABAA: {0}'.format(round(rosina_am_tim_baaa_abaa, 3)))\n",
    "print('*** Detection Metrics ***')\n",
    "print('D - p-value: {0} - {1}'.format(round(rosina_am_tim_d, 3), round(rosina_am_tim_pval_dict['d'], 3)))\n",
    "print('Danc - p-value: {0} - {1}'.format(round(rosina_am_tim_danc, 3), round(rosina_am_tim_pval_dict['danc'], 3)))\n",
    "print('D+ - p-value: {0} - {1}'.format(round(rosina_am_tim_dplus, 3), round(rosina_am_tim_pval_dict['dplus'], 3)))\n",
    "print('*** Quantification Metrics ***')\n",
    "print('fhom - std: {0} - {1}'.format(round(rosina_am_tim_fhom, 3), round(rosina_am_tim_bs_dict['fhom'], 3)))\n",
    "print('fanc - std: {0} - {1}'.format(round(rosina_am_tim_fanc, 3), round(rosina_am_tim_bs_dict['fanc'], 3)))\n",
    "print('f+ - std: {0} - {1}'.format(round(rosina_am_tim_fplus, 3), round(rosina_am_tim_bs_dict['fplus'], 3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introgression Between  _H. m. rosina_ & _H. cydno_\n",
    "Next, we re-examined rates of gene flow between _H. m. rosina_ and _H. cydno_ across two time periods: a short, recent period, and a long period.\n",
    "  \n",
    "First, we will look at the short time period 4—i.e., ((_H. m. melpomene (Panama)_, _H. m. rosina_), _H. cydno_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary for time period 4 ((H. m. melpomene (Panama), H. m. rosina), H. cydno)\n",
      "*** Site Pattern Counts ***\n",
      "ABBA - BABA: 4194.587\n",
      "BAAA - ABAA: -17724.129\n",
      "*** Detection Metrics ***\n",
      "D - p-value: 0.073 - 0.0\n",
      "Danc - p-value: -0.058 - 0.0\n",
      "D+ - p-value: -0.037 - 0.0\n",
      "*** Quantification Metrics ***\n",
      "fhom - std: 0.031 - 0.002\n",
      "fanc - std: -0.15 - 0.037\n",
      "f+ - std: -0.054 - 0.005\n"
     ]
    }
   ],
   "source": [
    "# Calculate site pattern counts for.\n",
    "melp_p_rosina_cydno_abba, melp_p_rosina_cydno_baba, melp_p_rosina_cydno_baaa, melp_p_rosina_cydno_abaa = butterfly_site_patterns(\n",
    "    p1=melp_p_der_freq_array,\n",
    "    p2=rosina_der_freq_array,\n",
    "    p3=cydno_der_freq_array,\n",
    ")\n",
    "melp_p_cydno_cydno_abba, melp_p_cydno_cydno_baba, melp_p_cydno_cydno_baaa, melp_p_cydno_cydno_abaa = butterfly_site_patterns(\n",
    "    p1=melp_p_der_freq_array,\n",
    "    p2=cydno_der_freq_array,\n",
    "    p3=cydno_der_freq_array,\n",
    ")\n",
    "# Calculate site pattern differences.\n",
    "melp_p_rosina_cydno_abba_baba = melp_p_rosina_cydno_abba - melp_p_rosina_cydno_baba\n",
    "melp_p_rosina_cydno_baaa_abaa = melp_p_rosina_cydno_baaa - melp_p_rosina_cydno_abaa\n",
    "# Calculate introgression detection metrics.\n",
    "melp_p_rosina_cydno_d, melp_p_rosina_cydno_danc, melp_p_rosina_cydno_dplus = butterfly_detection(\n",
    "    abba=melp_p_rosina_cydno_abba,\n",
    "    baba=melp_p_rosina_cydno_baba,\n",
    "    baaa=melp_p_rosina_cydno_baaa,\n",
    "    abaa=melp_p_rosina_cydno_abaa,\n",
    ")\n",
    "# Calculate introgression quantification metrics.\n",
    "melp_p_rosina_cydno_fhom, melp_p_rosina_cydno_fanc, melp_p_rosina_cydno_fplus = butterfly_quantification(\n",
    "    abba_num=melp_p_rosina_cydno_abba,\n",
    "    baba_num=melp_p_rosina_cydno_baba,\n",
    "    baaa_num=melp_p_rosina_cydno_baaa,\n",
    "    abaa_num=melp_p_rosina_cydno_abaa,\n",
    "    abba_den=ag_tim_tim_abba,\n",
    "    baba_den=ag_tim_tim_baba,\n",
    "    baaa_den=ag_tim_tim_baaa,\n",
    "    abaa_den=ag_tim_tim_abaa,\n",
    ")\n",
    "# Load in the bootstrapped distributions.\n",
    "melp_p_rosina_cydno_bs_dict = load_bs_stds(\n",
    "    p1='melP',\n",
    "    p2='ros',\n",
    "    p3='chi',\n",
    ")\n",
    "# Calculate p-values for introgression detection method.\n",
    "melp_p_rosina_cydno_pval_dict = calc_p_values(\n",
    "    obs_d=melp_p_rosina_cydno_d,\n",
    "    obs_danc=melp_p_rosina_cydno_danc,\n",
    "    obs_dplus=melp_p_rosina_cydno_dplus,\n",
    "    bs_std_d=melp_p_rosina_cydno_bs_dict['d'],\n",
    "    bs_std_danc=melp_p_rosina_cydno_bs_dict['danc'],\n",
    "    bs_std_dplus=melp_p_rosina_cydno_bs_dict['dplus'],\n",
    ")\n",
    "# Print a summary for this trio.\n",
    "print('Summary for time period 4 ((H. m. melpomene (Panama), H. m. rosina), H. cydno)')\n",
    "print('*** Site Pattern Counts ***')\n",
    "print('ABBA - BABA: {0}'.format(round(melp_p_rosina_cydno_abba_baba, 3)))\n",
    "print('BAAA - ABAA: {0}'.format(round(melp_p_rosina_cydno_baaa_abaa, 3)))\n",
    "print('*** Detection Metrics ***')\n",
    "print('D - p-value: {0} - {1}'.format(round(melp_p_rosina_cydno_d, 3), round(melp_p_rosina_cydno_pval_dict['d'], 3)))\n",
    "print('Danc - p-value: {0} - {1}'.format(round(melp_p_rosina_cydno_danc, 3), round(melp_p_rosina_cydno_pval_dict['danc'], 3)))\n",
    "print('D+ - p-value: {0} - {1}'.format(round(melp_p_rosina_cydno_dplus, 3), round(melp_p_rosina_cydno_pval_dict['dplus'], 3)))\n",
    "print('*** Quantification Metrics ***')\n",
    "print('fhom - std: {0} - {1}'.format(round(melp_p_rosina_cydno_fhom, 3), round(melp_p_rosina_cydno_bs_dict['fhom'], 3)))\n",
    "print('fanc - std: {0} - {1}'.format(round(melp_p_rosina_cydno_fanc, 3), round(melp_p_rosina_cydno_bs_dict['fanc'], 3)))\n",
    "print('f+ - std: {0} - {1}'.format(round(melp_p_rosina_cydno_fplus, 3), round(melp_p_rosina_cydno_bs_dict['fplus'], 3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will look at the long time period 4 + 3 + 2—i.e., ((_H. m. melpomene (French Guiana)_, _H. m. rosina_), _H. cydno_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary for time period 4 + 3 + 2 ((H. m. melpomene (French Guiana), H. m. rosina), H. cydno)\n",
      "*** Site Pattern Counts ***\n",
      "ABBA - BABA: 34801.297\n",
      "BAAA - ABAA: 66960.422\n",
      "*** Detection Metrics ***\n",
      "D - p-value: 0.493 - 0.0\n",
      "Danc - p-value: 0.147 - 0.0\n",
      "D+ - p-value: 0.194 - 0.0\n",
      "*** Quantification Metrics ***\n",
      "fhom - std: 0.258 - 0.007\n",
      "fanc - std: 0.568 - 0.008\n",
      "f+ - std: 0.403 - 0.006\n"
     ]
    }
   ],
   "source": [
    "# Calculate site pattern counts for.\n",
    "melp_fg_rosina_cydno_abba, melp_fg_rosina_cydno_baba, melp_fg_rosina_cydno_baaa, melp_fg_rosina_cydno_abaa = butterfly_site_patterns(\n",
    "    p1=melp_fg_der_freq_array,\n",
    "    p2=rosina_der_freq_array,\n",
    "    p3=cydno_der_freq_array,\n",
    ")\n",
    "melp_fg_cydno_cydno_abba, melp_fg_cydno_cydno_baba, melp_fg_cydno_cydno_baaa, melp_fg_cydno_cydno_abaa = butterfly_site_patterns(\n",
    "    p1=melp_fg_der_freq_array,\n",
    "    p2=cydno_der_freq_array,\n",
    "    p3=cydno_der_freq_array,\n",
    ")\n",
    "# Calculate site pattern differences.\n",
    "melp_fg_rosina_cydno_abba_baba = melp_fg_rosina_cydno_abba - melp_fg_rosina_cydno_baba\n",
    "melp_fg_rosina_cydno_baaa_abaa = melp_fg_rosina_cydno_baaa - melp_fg_rosina_cydno_abaa\n",
    "# Calculate introgression detection metrics.\n",
    "melp_fg_rosina_cydno_d, melp_fg_rosina_cydno_danc, melp_fg_rosina_cydno_dplus = butterfly_detection(\n",
    "    abba=melp_fg_rosina_cydno_abba,\n",
    "    baba=melp_fg_rosina_cydno_baba,\n",
    "    baaa=melp_fg_rosina_cydno_baaa,\n",
    "    abaa=melp_fg_rosina_cydno_abaa,\n",
    ")\n",
    "# Calculate introgression quantification metrics.\n",
    "melp_fg_rosina_cydno_fhom, melp_fg_rosina_cydno_fanc, melp_fg_rosina_cydno_fplus = butterfly_quantification(\n",
    "    abba_num=melp_fg_rosina_cydno_abba,\n",
    "    baba_num=melp_fg_rosina_cydno_baba,\n",
    "    baaa_num=melp_fg_rosina_cydno_baaa,\n",
    "    abaa_num=melp_fg_rosina_cydno_abaa,\n",
    "    abba_den=ag_tim_tim_abba,\n",
    "    baba_den=ag_tim_tim_baba,\n",
    "    baaa_den=ag_tim_tim_baaa,\n",
    "    abaa_den=ag_tim_tim_abaa,\n",
    ")\n",
    "# Load in the bootstrapped distributions.\n",
    "melp_fg_rosina_cydno_bs_dict = load_bs_stds(\n",
    "    p1='melFG',\n",
    "    p2='ros',\n",
    "    p3='chi',\n",
    ")\n",
    "# Calculate p-values for introgression detection method.\n",
    "melp_fg_rosina_cydno_pval_dict = calc_p_values(\n",
    "    obs_d=melp_fg_rosina_cydno_d,\n",
    "    obs_danc=melp_fg_rosina_cydno_danc,\n",
    "    obs_dplus=melp_fg_rosina_cydno_dplus,\n",
    "    bs_std_d=melp_fg_rosina_cydno_bs_dict['d'],\n",
    "    bs_std_danc=melp_fg_rosina_cydno_bs_dict['danc'],\n",
    "    bs_std_dplus=melp_fg_rosina_cydno_bs_dict['dplus'],\n",
    ")\n",
    "# Print a summary for this trio.\n",
    "print('Summary for time period 4 + 3 + 2 ((H. m. melpomene (French Guiana), H. m. rosina), H. cydno)')\n",
    "print('*** Site Pattern Counts ***')\n",
    "print('ABBA - BABA: {0}'.format(round(melp_fg_rosina_cydno_abba_baba, 3)))\n",
    "print('BAAA - ABAA: {0}'.format(round(melp_fg_rosina_cydno_baaa_abaa, 3)))\n",
    "print('*** Detection Metrics ***')\n",
    "print('D - p-value: {0} - {1}'.format(round(melp_fg_rosina_cydno_d, 3), round(melp_fg_rosina_cydno_pval_dict['d'], 3)))\n",
    "print('Danc - p-value: {0} - {1}'.format(round(melp_fg_rosina_cydno_danc, 3), round(melp_fg_rosina_cydno_pval_dict['danc'], 3)))\n",
    "print('D+ - p-value: {0} - {1}'.format(round(melp_fg_rosina_cydno_dplus, 3), round(melp_fg_rosina_cydno_pval_dict['dplus'], 3)))\n",
    "print('*** Quantification Metrics ***')\n",
    "print('fhom - std: {0} - {1}'.format(round(melp_fg_rosina_cydno_fhom, 3), round(melp_fg_rosina_cydno_bs_dict['fhom'], 3)))\n",
    "print('fanc - std: {0} - {1}'.format(round(melp_fg_rosina_cydno_fanc, 3), round(melp_fg_rosina_cydno_bs_dict['fanc'], 3)))\n",
    "print('f+ - std: {0} - {1}'.format(round(melp_fg_rosina_cydno_fplus, 3), round(melp_fg_rosina_cydno_bs_dict['fplus'], 3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will look at the long time period 4 + 3 + 2—i.e., ((_H. m. amaryllis_, _H. m. rosina_), _H. cydno_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary for time period 4 + 3 + 2 ((H. m. amaryllis, H. m. rosina), H. cydno)\n",
      "*** Site Pattern Counts ***\n",
      "ABBA - BABA: 37885.734\n",
      "BAAA - ABAA: 58260.551\n",
      "*** Detection Metrics ***\n",
      "D - p-value: 0.49 - 0.0\n",
      "Danc - p-value: 0.127 - 0.0\n",
      "D+ - p-value: 0.179 - 0.0\n",
      "*** Quantification Metrics ***\n",
      "fhom - std: 0.281 - 0.007\n",
      "fanc - std: 0.494 - 0.008\n",
      "f+ - std: 0.38 - 0.007\n"
     ]
    }
   ],
   "source": [
    "# Calculate site pattern counts for.\n",
    "amaryllis_rosina_cydno_abba, amaryllis_rosina_cydno_baba, amaryllis_rosina_cydno_baaa, amaryllis_rosina_cydno_abaa = butterfly_site_patterns(\n",
    "    p1=amaryllis_der_freq_array,\n",
    "    p2=rosina_der_freq_array,\n",
    "    p3=cydno_der_freq_array,\n",
    ")\n",
    "amaryllis_cydno_cydno_abba, amaryllis_cydno_cydno_baba, amaryllis_cydno_cydno_baaa, amaryllis_cydno_cydno_abaa = butterfly_site_patterns(\n",
    "    p1=amaryllis_der_freq_array,\n",
    "    p2=cydno_der_freq_array,\n",
    "    p3=cydno_der_freq_array,\n",
    ")\n",
    "# Calculate site pattern differences.\n",
    "amaryllis_rosina_cydno_abba_baba = amaryllis_rosina_cydno_abba - amaryllis_rosina_cydno_baba\n",
    "amaryllis_rosina_cydno_baaa_abaa = amaryllis_rosina_cydno_baaa - amaryllis_rosina_cydno_abaa\n",
    "# Calculate introgression detection metrics.\n",
    "amaryllis_rosina_cydno_d, amaryllis_rosina_cydno_danc, amaryllis_rosina_cydno_dplus = butterfly_detection(\n",
    "    abba=amaryllis_rosina_cydno_abba,\n",
    "    baba=amaryllis_rosina_cydno_baba,\n",
    "    baaa=amaryllis_rosina_cydno_baaa,\n",
    "    abaa=amaryllis_rosina_cydno_abaa,\n",
    ")\n",
    "# Calculate introgression quantification metrics.\n",
    "amaryllis_rosina_cydno_fhom, amaryllis_rosina_cydno_fanc, amaryllis_rosina_cydno_fplus = butterfly_quantification(\n",
    "    abba_num=amaryllis_rosina_cydno_abba,\n",
    "    baba_num=amaryllis_rosina_cydno_baba,\n",
    "    baaa_num=amaryllis_rosina_cydno_baaa,\n",
    "    abaa_num=amaryllis_rosina_cydno_abaa,\n",
    "    abba_den=ag_tim_tim_abba,\n",
    "    baba_den=ag_tim_tim_baba,\n",
    "    baaa_den=ag_tim_tim_baaa,\n",
    "    abaa_den=ag_tim_tim_abaa,\n",
    ")\n",
    "# Load in the bootstrapped distributions.\n",
    "amaryllis_rosina_cydno_bs_dict = load_bs_stds(\n",
    "    p1='am',\n",
    "    p2='ros',\n",
    "    p3='chi',\n",
    ")\n",
    "# Calculate p-values for introgression detection method.\n",
    "amaryllis_rosina_cydno_pval_dict = calc_p_values(\n",
    "    obs_d=amaryllis_rosina_cydno_d,\n",
    "    obs_danc=amaryllis_rosina_cydno_danc,\n",
    "    obs_dplus=amaryllis_rosina_cydno_dplus,\n",
    "    bs_std_d=amaryllis_rosina_cydno_bs_dict['d'],\n",
    "    bs_std_danc=amaryllis_rosina_cydno_bs_dict['danc'],\n",
    "    bs_std_dplus=amaryllis_rosina_cydno_bs_dict['dplus'],\n",
    ")\n",
    "# Print a summary for this trio.\n",
    "print('Summary for time period 4 + 3 + 2 ((H. m. amaryllis, H. m. rosina), H. cydno)')\n",
    "print('*** Site Pattern Counts ***')\n",
    "print('ABBA - BABA: {0}'.format(round(amaryllis_rosina_cydno_abba_baba, 3)))\n",
    "print('BAAA - ABAA: {0}'.format(round(amaryllis_rosina_cydno_baaa_abaa, 3)))\n",
    "print('*** Detection Metrics ***')\n",
    "print('D - p-value: {0} - {1}'.format(round(amaryllis_rosina_cydno_d, 3), round(amaryllis_rosina_cydno_pval_dict['d'], 3)))\n",
    "print('Danc - p-value: {0} - {1}'.format(round(amaryllis_rosina_cydno_danc, 3), round(amaryllis_rosina_cydno_pval_dict['danc'], 3)))\n",
    "print('D+ - p-value: {0} - {1}'.format(round(amaryllis_rosina_cydno_dplus, 3), round(amaryllis_rosina_cydno_pval_dict['dplus'], 3)))\n",
    "print('*** Quantification Metrics ***')\n",
    "print('fhom - std: {0} - {1}'.format(round(amaryllis_rosina_cydno_fhom, 3), round(amaryllis_rosina_cydno_bs_dict['fhom'], 3)))\n",
    "print('fanc - std: {0} - {1}'.format(round(amaryllis_rosina_cydno_fanc, 3), round(amaryllis_rosina_cydno_bs_dict['fanc'], 3)))\n",
    "print('f+ - std: {0} - {1}'.format(round(amaryllis_rosina_cydno_fplus, 3), round(amaryllis_rosina_cydno_bs_dict['fplus'], 3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lastly, we will look at the long time period 4 + 3 + 2—i.e., ((_H. m. aglaope_, _H. m. rosina_), _H. cydno_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary for time period 4 + 3 + 2 ((H. m. aglaope, H. m. rosina), H. cydno)\n",
      "*** Site Pattern Counts ***\n",
      "ABBA - BABA: 38524.203\n",
      "BAAA - ABAA: 57445.078\n",
      "*** Detection Metrics ***\n",
      "D - p-value: 0.501 - 0.0\n",
      "Danc - p-value: 0.126 - 0.0\n",
      "D+ - p-value: 0.18 - 0.0\n",
      "*** Quantification Metrics ***\n",
      "fhom - std: 0.286 - 0.007\n",
      "fanc - std: 0.487 - 0.009\n",
      "f+ - std: 0.38 - 0.007\n"
     ]
    }
   ],
   "source": [
    "# Calculate site pattern counts for.\n",
    "aglaope_rosina_cydno_abba, aglaope_rosina_cydno_baba, aglaope_rosina_cydno_baaa, aglaope_rosina_cydno_abaa = butterfly_site_patterns(\n",
    "    p1=aglaope_der_freq_array,\n",
    "    p2=rosina_der_freq_array,\n",
    "    p3=cydno_der_freq_array,\n",
    ")\n",
    "aglaope_cydno_cydno_abba, aglaope_cydno_cydno_baba, aglaope_cydno_cydno_baaa, aglaope_cydno_cydno_abaa = butterfly_site_patterns(\n",
    "    p1=aglaope_der_freq_array,\n",
    "    p2=cydno_der_freq_array,\n",
    "    p3=cydno_der_freq_array,\n",
    ")\n",
    "# Calculate site pattern differences.\n",
    "aglaope_rosina_cydno_abba_baba = aglaope_rosina_cydno_abba - aglaope_rosina_cydno_baba\n",
    "aglaope_rosina_cydno_baaa_abaa = aglaope_rosina_cydno_baaa - aglaope_rosina_cydno_abaa\n",
    "# Calculate introgression detection metrics.\n",
    "aglaope_rosina_cydno_d, aglaope_rosina_cydno_danc, aglaope_rosina_cydno_dplus = butterfly_detection(\n",
    "    abba=aglaope_rosina_cydno_abba,\n",
    "    baba=aglaope_rosina_cydno_baba,\n",
    "    baaa=aglaope_rosina_cydno_baaa,\n",
    "    abaa=aglaope_rosina_cydno_abaa,\n",
    ")\n",
    "# Calculate introgression quantification metrics.\n",
    "aglaope_rosina_cydno_fhom, aglaope_rosina_cydno_fanc, aglaope_rosina_cydno_fplus = butterfly_quantification(\n",
    "    abba_num=aglaope_rosina_cydno_abba,\n",
    "    baba_num=aglaope_rosina_cydno_baba,\n",
    "    baaa_num=aglaope_rosina_cydno_baaa,\n",
    "    abaa_num=aglaope_rosina_cydno_abaa,\n",
    "    abba_den=ag_tim_tim_abba,\n",
    "    baba_den=ag_tim_tim_baba,\n",
    "    baaa_den=ag_tim_tim_baaa,\n",
    "    abaa_den=ag_tim_tim_abaa,\n",
    ")\n",
    "# Load in the bootstrapped distributions.\n",
    "aglaope_rosina_cydno_bs_dict = load_bs_stds(\n",
    "    p1='ag',\n",
    "    p2='ros',\n",
    "    p3='chi',\n",
    ")\n",
    "# Calculate p-values for introgression detection method.\n",
    "aglaope_rosina_cydno_pval_dict = calc_p_values(\n",
    "    obs_d=aglaope_rosina_cydno_d,\n",
    "    obs_danc=aglaope_rosina_cydno_danc,\n",
    "    obs_dplus=aglaope_rosina_cydno_dplus,\n",
    "    bs_std_d=aglaope_rosina_cydno_bs_dict['d'],\n",
    "    bs_std_danc=aglaope_rosina_cydno_bs_dict['danc'],\n",
    "    bs_std_dplus=aglaope_rosina_cydno_bs_dict['dplus'],\n",
    ")\n",
    "# Print a summary for this trio.\n",
    "print('Summary for time period 4 + 3 + 2 ((H. m. aglaope, H. m. rosina), H. cydno)')\n",
    "print('*** Site Pattern Counts ***')\n",
    "print('ABBA - BABA: {0}'.format(round(aglaope_rosina_cydno_abba_baba, 3)))\n",
    "print('BAAA - ABAA: {0}'.format(round(aglaope_rosina_cydno_baaa_abaa, 3)))\n",
    "print('*** Detection Metrics ***')\n",
    "print('D - p-value: {0} - {1}'.format(round(aglaope_rosina_cydno_d, 3), round(aglaope_rosina_cydno_pval_dict['d'], 3)))\n",
    "print('Danc - p-value: {0} - {1}'.format(round(aglaope_rosina_cydno_danc, 3), round(aglaope_rosina_cydno_pval_dict['danc'], 3)))\n",
    "print('D+ - p-value: {0} - {1}'.format(round(aglaope_rosina_cydno_dplus, 3), round(aglaope_rosina_cydno_pval_dict['dplus'], 3)))\n",
    "print('*** Quantification Metrics ***')\n",
    "print('fhom - std: {0} - {1}'.format(round(aglaope_rosina_cydno_fhom, 3), round(aglaope_rosina_cydno_bs_dict['fhom'], 3)))\n",
    "print('fanc - std: {0} - {1}'.format(round(aglaope_rosina_cydno_fanc, 3), round(aglaope_rosina_cydno_bs_dict['fanc'], 3)))\n",
    "print('f+ - std: {0} - {1}'.format(round(aglaope_rosina_cydno_fplus, 3), round(aglaope_rosina_cydno_bs_dict['fplus'], 3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bootstrapping example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import time\n",
    "def butterfly_bootstrap(\n",
    "    genome_df,\n",
    "    p1_id,\n",
    "    p2_id,\n",
    "    p3_id,\n",
    "    n_replicates,\n",
    "):\n",
    "    \"\"\"\n",
    "    ###########################################################################\n",
    "    INPUT\n",
    "        genome_df: Whole-genome dataframe for butterflies.\n",
    "        p1_id: ID for the P1 population.\n",
    "        p2_id: ID for the P2 population.\n",
    "        p3_id: ID for the P3 population.\n",
    "        n_replicates: Number of bootstrap replicates.\n",
    "    ---------------------------------------------------------------------------\n",
    "    OUTPUT: List of site pattern counts from bootstrapped replicates.\n",
    "    ###########################################################################\n",
    "    \"\"\"\n",
    "    # Intialize arrays to store bootstrapped values.\n",
    "    abba_array = np.array([])\n",
    "    baba_array = np.array([])\n",
    "    baaa_array = np.array([])\n",
    "    abaa_array = np.array([])\n",
    "    abba_hom_array = np.array([])\n",
    "    baba_hom_array = np.array([])\n",
    "    baaa_hom_array = np.array([])\n",
    "    abaa_hom_array = np.array([])\n",
    "    # Intialize a dictionary of chromosome lengths from table S4.6.1 (doi:10.1038/nature11041).\n",
    "    chromosome_dicc = {\n",
    "        'chr1': 15_755_848, 'chr2': 3_590_614, 'chr3': 8_943_778,\n",
    "        'chr4': 6_670_749, 'chr5': 8_000_463, 'chr6': 13_155_959,\n",
    "        'chr7': 11_792_147, 'chr8': 6_828_437, 'chr9': 8_311_378,\n",
    "        'chr10': 17_492_759, 'chr11': 11_233_900, 'chr12': 15_835_627,\n",
    "        'chr13': 13_526_828, 'chr14': 6_527_076, 'chr15': 7_742_867,\n",
    "        'chr16': 9_269_283, 'chr17': 13_935_249, 'chr18': 15_453_272,\n",
    "        'chr19': 14_778_520, 'chr20': 5_748_428, 'chrZ': 4_088_377,\n",
    "    }\n",
    "    # For each bootstrap replicate.\n",
    "    for i in range(n_replicates):\n",
    "        # Intialize empty arrays to store bootstrapped genomes.\n",
    "        bootstrapped_p1 = np.array([])\n",
    "        bootstrapped_p2 = np.array([])\n",
    "        bootstrapped_p3 = np.array([])\n",
    "        # For each bootstrap window (219 x 1 Mb).\n",
    "        for j in range(219):\n",
    "            # Randomly select a chromosome.\n",
    "            chromosome = random.choice(list(chromosome_dicc.keys()))\n",
    "            # Randomly generate a start and end position.\n",
    "            start = np.random.randint((chromosome_dicc[chromosome] - 999_999))\n",
    "            end = start + 1_000_000\n",
    "            # Subset the whole-genome dataframe to only include the randomly selected chromosome.\n",
    "            chromosome_df = genome_df[genome_df['chromosome'] == chromosome]\n",
    "            # Extract the variable positions for that chromosome.\n",
    "            variable_positions = chromosome_df['chrom_pos'].values\n",
    "            # Identify the variants that fall within the 1 Mb window.\n",
    "            variants = np.where(((start <= variable_positions) & (variable_positions <= end)))[0]\n",
    "            # Extract the focal populations to subset.\n",
    "            p1_chromosome = chromosome_df[p1_id].values\n",
    "            p2_chromosome = chromosome_df[p2_id].values\n",
    "            p3_chromosome = chromosome_df[p3_id].values\n",
    "            # Subset the 1 Mb window from each focal population.\n",
    "            p1_subset = p1_chromosome[variants]\n",
    "            p2_subset = p2_chromosome[variants]\n",
    "            p3_subset = p3_chromosome[variants]\n",
    "            # Append 219 windows of 1 Mb to make bootstrapped genomes.\n",
    "            bootstrapped_p1 = np.append(bootstrapped_p1, p1_subset)\n",
    "            bootstrapped_p2 = np.append(bootstrapped_p2, p2_subset)\n",
    "            bootstrapped_p3 = np.append(bootstrapped_p3, p3_subset)\n",
    "        # Calculate site pattern counts for the bootstrapped replicate and store the results.\n",
    "        abba, baba, baaa, abaa = butterfly_site_patterns(\n",
    "            p1=bootstrapped_p1,\n",
    "            p2=bootstrapped_p2,\n",
    "            p3=bootstrapped_p3,\n",
    "        )\n",
    "        abba_array = np.append(abba_array, abba)\n",
    "        baba_array = np.append(baba_array, baba)\n",
    "        baaa_array = np.append(baaa_array, baaa)\n",
    "        abaa_array = np.append(abaa_array, abaa)\n",
    "        abba_hom, baba_hom, baaa_hom, abaa_hom = butterfly_site_patterns(\n",
    "            p1=bootstrapped_p1,\n",
    "            p2=bootstrapped_p3,\n",
    "            p3=bootstrapped_p3,\n",
    "        )\n",
    "        abba_hom_array = np.append(abba_hom_array, abba_hom)\n",
    "        baba_hom_array = np.append(baba_hom_array, baba_hom)\n",
    "        baaa_hom_array = np.append(baaa_hom_array, baaa_hom)\n",
    "        abaa_hom_array = np.append(abaa_hom_array, abaa_hom)\n",
    "    return abba_array, baba_array, baaa_array, abaa_array, abba_hom_array, baba_hom_array, baaa_hom_array, abaa_hom_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run Time: 0.816 minutes\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "bs_abba, bs_baba, bs_baaa, bs_abaa, bs_abba_hom, bs_baba_hom, bs_baaa_hom, bs_abaa_hom = butterfly_bootstrap(\n",
    "    genome_df=der_freq_df,\n",
    "    p1_id='ag',\n",
    "    p2_id='ros',\n",
    "    p3_id='chi',\n",
    "    n_replicates=1,\n",
    ")\n",
    "end = time.time()\n",
    "print('Run Time: {} minutes'.format(round(((end - start) / 60), 3)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "abba: replicates = 1; average = 50249.869\n",
      "baba: replicates = 1; average = 17282.744\n",
      "baaa: replicates = 1; average = 236909.162\n",
      "abaa: replicates = 1; average = 183353.037\n",
      "abba hom: replicates = 1; average = 147773.385\n",
      "baba hom: replicates = 1; average = 11446.041\n",
      "baaa hom: replicates = 1; average = 271200.443\n",
      "abaa hom: replicates = 1; average = 171771.35\n"
     ]
    }
   ],
   "source": [
    "# Print out a summary of each distribution.\n",
    "print('abba: replicates = {0}; average = {1}'.format(bs_abba.shape[0], round(np.mean(bs_abba), 3)))\n",
    "print('baba: replicates = {0}; average = {1}'.format(bs_baba.shape[0], round(np.mean(bs_baba), 3)))\n",
    "print('baaa: replicates = {0}; average = {1}'.format(bs_baaa.shape[0], round(np.mean(bs_baaa), 3)))\n",
    "print('abaa: replicates = {0}; average = {1}'.format(bs_abaa.shape[0], round(np.mean(bs_abaa), 3)))\n",
    "print('abba hom: replicates = {0}; average = {1}'.format(bs_abba_hom.shape[0], round(np.mean(bs_abba_hom), 3)))\n",
    "print('baba hom: replicates = {0}; average = {1}'.format(bs_baba_hom.shape[0], round(np.mean(bs_baba_hom), 3)))\n",
    "print('baaa hom: replicates = {0}; average = {1}'.format(bs_baaa_hom.shape[0], round(np.mean(bs_baaa_hom), 3)))\n",
    "print('abaa hom: replicates = {0}; average = {1}'.format(bs_abaa_hom.shape[0], round(np.mean(bs_abaa_hom), 3)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anc_der_intro_env",
   "language": "python",
   "name": "anc_der_intro_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
